### ISSUE TITLE
AKPlayer.load() on a file crashes after upgrading to 4.10 (from 4.9.4)

### ISSUE THREAD

**Issue by ghubaddole** [AudioKit/AudioKit#2073]
AudioKit.AKFaderAudioUnit input format must match output format

2020-06-02 12:15:08.494929+0530  [avae]            AVAEInternal.h:109   [AVAudioEngineGraph.mm:4170:UpdateGraphAfterReconfig: (AUGraphParser::InitializeActiveNodesInOutputChain(ThisGraph, kOutputChainFullTraversal, *conn.srcNode, isChainActive)): error -10868
2020-06-02 12:15:08.497253+0530 *** Terminating app due to uncaught exception 'com.apple.coreaudio.avfaudio', reason: 'error -10868'

**Comment by ghubaddole** [comment]
Another observation - it happens when connected to a bluetooth speaker with a different sample rate than the file - so, my  file is 48K, bluetooth is 44.1K. If I disconnect bluetooth, the crash goes away.

**Comment by ghubaddole** [comment]
Could the recent AKSettings sampleRate / audioFormat changes be causing the issue?

I have a file, 48 KHz, 32 bits/sample (default recording format on iphone x). 
I connect to a audio route that has a 44.1 Khz sample rate.
then, a player.load(url: "path to fie") crashes.
If I disconnect the audio route, the app needs a kill, before the native sample rate is applied. a restarted app on the device's default speaker works fine.

**Comment by ghubaddole** [comment]
Yet another clue -- I went back to 4.9.4 and this crash is not happen anymore ... however, if I use
player.fadeOutAndStop(time: 1.0), upon stopping player, there's the same crash (this is on the device, native sample rate). Funnily, I was not using fadeOutAndStop in 4.10, and the crash happened with a speaker of different sample rate. 

-gd

**Comment by ryanfrancesconi** [comment]
Just to clarify, you are playing back a stereo 48k file and AudioKit.output is set to 44.1k, correct?

**Comment by ryanfrancesconi** [comment]
Work around for the moment is to just make a new AKPlayer if you are changing the sample rate:

 // for now just make a new player so it's not necessary to rebalance format types
// if the processingFormat changes
        if player != nil {
            player?.detach()
            player = nil
        }
        player = AKDynamicPlayer(url: url)

// plug it back in

**Comment by ryanfrancesconi** [comment]
Hi, I've just pushed up an update that will address this. This will be in the v5 develop branch for now, but you could handle on your end as well by monitoring processingFormats. 

For performance reasons, sample rate isn't quite as flexible as it used to be. Only if necessary will AKPlayer do real time sample rate conversion and this requires installing an AKMixer into the graph. This means, that if you want to change the processingFormat, you should create a new AKPlayer and detach the old one. This was unclear previously with that load() command, and I've now just pushed up a version that will throw and suggest you make a new player in this scenario.

I've also updated the macOSDevelopment project to show this in practice:

```
    /// open an audio URL for playing
    func open(url: URL) {
        inputSourceInfo.stringValue = url.lastPathComponent

        do {
            try player?.load(url: url)

        } catch let err as NSError {
            AKLog("🚩", err.description)
            createPlayer(url: url)
        }

        AKLog("Opened", url.path, "duration", player?.duration, "processingFormat", player?.processingFormat)
    }

    private func createPlayer(url: URL) {
        // make a new player so it's not necessary to rebalance format types
        // if the processingFormat changes
        if player != nil {
            player?.detach()
            player = nil
        }
        AKLog("Creating player...", url)
        player = AKDynamicPlayer(url: url)
        player?.completionHandler = handleAudioComplete
        player?.isLooping = loopButton.state == .on
        connectPlayer()
    }

```
Will likely be merged into v5 today.

**Comment by ghubaddole** [comment]
Thanks a ton, Ryan!
I've been unclear on this before and avoiding reconnecting notes dynamically like the plague... 
Can you advice as to how to reattach notes in an existing (rather complex) graph without restarting the engine?
I'm sure it'll help other devs too, based on my searches on SO and audiokit issues, etc.
Thank you in advance.

**Comment by ghubaddole** [comment]
Thanks for the detailed reply!! Really appreciate all your help :-)
Is there a way to stave off any format / route changes, for example in the middle of renderToFile? That seems to truncate recordings (at least thats the theory I have)

**Comment by ryanfrancesconi** [comment]
sorry, i don't really work with iOS much, so I don't have any good advice for those types of changes, if that's what you mean.

It seems to me that the application shouldn't allow format changes during rendering.

**Comment by ghubaddole** [comment]
Any chance the fix will be in v4? v5 is a few months away (I tried but my code isn't quite working. since there's a few months before it's ready to be tested, there'll be a lot of changes between now and then, so will try later).

**Comment by ryanfrancesconi** [comment]
i think we can merge that change to v4 yes... @aure  ?

**Comment by ryanfrancesconi** [comment]
Made the change in v4, though, it is a fairly minor difference, You'll still need to catch the error on your load() and rebuild your player graph if the format changes like the previous example I gave

**Comment by ghubaddole** [comment]
Many thanks! Yes noted -- I'm going to need to restructure code quite a bit to allow for rebuilding the graph in the middle of the app.

**Comment by ryanfrancesconi** [comment]
it really shouldn't be too hard to do as a player is generally a node at the end of a chain. a player doesn't accept input, so it just needs to be detached and reattached.

**Comment by quantamrhino** [comment]
Aloha! 
I have a simple music player app that creates a chain on startup (player -> eq -> limiter -> mixer -> output).
I have a list of tracks and when a track is selected, I load the corresponding file to the player. when the user navigates back, i stop playing.
I've been using the latest v4 branch and now the v4 dev branch, and running into issues that could be potentially related to this bug perhaps.

I have a simple app to demo the problem.
4 buttons : play 44.1K file, play 48K file, play a oscillator tone, stop all.
On all sample rates (on simulator @ 44.1K and physical devices with 48K rate), oscillator works fine.

On my iphone 7, (48K native sample rate), I set the a 48K rate and then the  play48() and playTone() works. but play44() crashes with the following:

2020-06-22 18:56:33.265543+0530 rhinoPlayer[37426:1843219] AudioKit.AKFaderAudioUnit input format must match output format

Even if I set the audiokit sample rate to 44.1K before playing, it's the same crash. 

Here's the project, I'm building from audiokit v4 developer branch.
https://drive.google.com/drive/folders/1Gg218GAtefB__rAvR3mMTqPzC5MNOjuX?usp=sharing 

I'm not sure if i'm doing something wrong / not doing something or if it's a genuine bug.
Also, in this kind of chain, player is not a child node. how can I change the player without having a crash? I'd really appreciate some advice!

Aloha!

**Comment by ryanfrancesconi** [comment]
it sounds like your processing formats don't match. 

if they can't, then put an AKMixer before the player so that sample rate conversion happens. I don't recommend that though.

AKSettings must be set before you establish any connections whatsoever.

**Comment by quantamrhino** [comment]
Thanks Ryan,
I have one file that's wav at 44.1K, another that's wav at 48K. Users can import files of various formats into the app (in theory).

Why is a mixer before player not recommended?
What's the downside? To clarify, do you mean:
mixer(no inputs) -> player -> etc -> out? How will loading a file to player that's a different format affect the mixer above? Not sure I grok.

I could have files of different formats. How do I handle them at run time then?

**Comment by quantamrhino** [comment]
since player can't take a input node, not sure how to connect a mixer before. I see you've deleted that part of the comment though.
The thing is that this was working before, on 4.9.4. now it's crashing reliably. I don't see how I can update to further versions -- looks like i may need to pull my app from the app store at some point in the future unless I can figure this out :-(

**Comment by ryanfrancesconi** [comment]
you just need to look at the format of your incoming file. If it is different, then you should make a new AKPlayer and reconnect it. This is pretty simple to do as a player is the sound source.

AKPlayer will install a mixer on instantiation if formats differ. It won't on a load(). I can consider adding that though. Note, I don't recommend this approach due to performance reasons. Realtime sample rate conversion isn't ideal.

You can of course, just make your own player too. The source code is all there.

**Comment by ryanfrancesconi** [comment]
An example of that load() with a new player looks like is in the macOSDevelopment project:

https://github.com/AudioKit/AudioKit/blob/v5-develop/Developer/macOS/macOSDevelopment/macOSDevelopment/ViewController.swift

See: open(url:

**Comment by quantamrhino** [comment]
Aah i see! This code makes it very clear! Thanks a ton, Ryan!
Aloha!

**Comment by quantamrhino** [comment]
Aloha!
I think this is what i've been hitting! (https://github.com/AudioKit/AudioKit/pull/2131)
Thanks for fixing this :-)
Any change this can be put in v4-dev? I'd love to start using v5 but my code crashes. Looks like v5 isn't ready for primetime yet...?

**Comment by ryanfrancesconi** [comment]
i think we did merge to v4...

**Comment by ryanfrancesconi** [comment]
oh you mean the fade out one? could prob merge that i think

**Comment by ryanfrancesconi** [comment]
It would be more useful to identify what is crashing for you in AK5 though?

**Comment by quantamrhino** [comment]
> oh you mean the fade out one? could prob merge that i think

That'll be great. I'm 99.99% sure that's what I'm hitting in 4.10. That should unblock my app!

> It would be more useful to identify what is crashing for you in AK5 though?

I tried to migrate to 5dev - took time (Doubles to AUValue all over, parameter -> parameter.value, AudioKit->AKManager etc).
At the end it crashed when setting up the graph. Assuming the branch isn't ready. I reverted. :-( Should have saved a copy.
If you think I can port my production app without a great deal of disruption to v5 dev, I'd be happy to. Or, when you think its ready - say in a month or so?

**Comment by ryanfrancesconi** [comment]
I am using v5 in a production setting, so of course am interested to hear exactly which parts are unstable in other areas. It's going ok for me on macOS. But, we're not really going to be supporting v4 builds at some point which is up to @aure 

So, it's best to try to migrate to v5 as soon as you can.

Yeah, all the AUValue vs Double is a pain, but hopefully not too much else.

**Comment by quantamrhino** [comment]
https://github.com/AudioKit/AudioKit/issues/2134#issuecomment-654186903 beat me to it.
I'm seeing the same problem with AKSampler.
Aloha
-Q
